lib/MySQL_Thread.cpp

```diff
@@ -3558,90 +3271,35 @@ MySQL_Session * MySQL_Thread::create_new_session_and_client_data_stream(int _fd)
-	sess->client_myds->myconn->options.sql_safe_updates_int = sql_safe_updates_int;
-	if (sess->client_myds->myconn->options.sql_safe_updates) {
-		free(sess->client_myds->myconn->options.sql_safe_updates);
-	}
-	sess->client_myds->myconn->options.sql_safe_updates=strdup(mysql_thread___default_sql_safe_updates);
-
-	uint32_t collation_connection_int=SpookyHash::Hash32(mysql_thread___default_collation_connection,strlen(mysql_thread___default_collation_connection),10);
-	sess->client_myds->myconn->options.collation_connection_int = collation_connection_int;
-	if (sess->client_myds->myconn->options.collation_connection) {
-		free(sess->client_myds->myconn->options.collation_connection);
-	}
-	sess->client_myds->myconn->options.collation_connection=strdup(mysql_thread___default_collation_connection);
-
-	uint32_t net_write_timeout_int=SpookyHash::Hash32(mysql_thread___default_net_write_timeout,strlen(mysql_thread___default_net_write_timeout),10);
-	sess->client_myds->myconn->options.net_write_timeout_int = net_write_timeout_int;
-	if (sess->client_myds->myconn->options.net_write_timeout) {
-		free(sess->client_myds->myconn->options.net_write_timeout);
-	}
-	sess->client_myds->myconn->options.net_write_timeout=strdup(mysql_thread___default_net_write_timeout);
-
-	uint32_t max_join_size_int=SpookyHash::Hash32(mysql_thread___default_max_join_size,strlen(mysql_thread___default_max_join_size),10);
-	sess->client_myds->myconn->options.max_join_size_int = max_join_size_int;
-	if (sess->client_myds->myconn->options.max_join_size) {
-		free(sess->client_myds->myconn->options.max_join_size);
+	for (int i=0; i<SQL_NAME_LAST; i++) {
+		if (i == SQL_CHARACTER_SET || i == SQL_CHARACTER_SET_RESULTS ||
+			i == SQL_CHARACTER_SET_CONNECTION || i == SQL_CHARACTER_SET_CLIENT ||
+			i == SQL_CHARACTER_SET_DATABASE) {
+			const MARIADB_CHARSET_INFO *ci = NULL;
+			ci = proxysql_find_charset_name(mysql_thread___default_variables[i]);
+			if (!ci) {
+				proxy_error("Cannot find character set for name [%s]. Configuration error. Check [%s] global variable. Using character set 33.\n", 
+						mysql_thread___default_variables[i], mysql_tracked_variables[i].internal_variable_name);
+				assert(0);
+			}
+			std::stringstream ss;
+			ss << ci->nr;
+			sess->mysql_variables->client_set_value(i, ss.str());
+		} else if (i == SQL_COLLATION_CONNECTION) {
+			const MARIADB_CHARSET_INFO *ci = NULL;
+			ci = proxysql_find_charset_collate(mysql_thread___default_variables[i]);
+			if (!ci) {
+				proxy_error("Cannot find character set for name [%s]. Configuration error. Check [%s] global variable. Using character set 33.\n", 
+						mysql_thread___default_variables[SQL_COLLATION_CONNECTION], mysql_tracked_variables[SQL_COLLATION_CONNECTION].internal_variable_name);
+				assert(0);
+			}
+			std::stringstream ss;
+			ss << ci->nr;
+			sess->mysql_variables->client_set_value(i, ss.str());
+		} else {
+			sess->mysql_variables->client_set_value(i, mysql_thread___default_variables[i]);
```

Comment:
For now, I am not sure about this.
If the client didn't set the variable, and a default it is not required (for example, `collation_connection`), probably we should not assign these variables.
To me, assigning a variable means that it needs to be tracked.

I may remove this comment after further review.

Reply:
I have removed this code completely. Client character set initialization takes place in 
```
 bool MySQL_Protocol::process_pkt_handshake_response(unsigned char *pkt, unsigned int len)
```
From my understanding, default values, that are configured in the `global_variables` should not be used to initialize character set variables.

Reply:
Correct.
As part of the handshake, in `MySQL_Protocol::generate_pkt_initial_handshake()` , we have this:
```
  const MARIADB_CHARSET_INFO *ci = NULL;
  ci = proxysql_find_charset_name(mysql_thread___default_variables[SQL_CHARACTER_SET]);
  if (!ci) {
      proxy_error("Cannot find character set for name [%s]. Configuration error. Check [%s] global variable.\n",
              mysql_thread___default_variables[SQL_CHARACTER_SET], mysql_tracked_variables[SQL_CHARACTER_SET].internal_variable_name);
      assert(0);
  }
  uint8_t uint8_charset = ci->nr & 255;
```
The client can, in `process_pkt_handshake_response`, either:
* accept it replying with the same collation id
* accept it replying with value 0 see #810
* change it returning a different collation